// jsonresume-theme-gnap/helpers.js
var Handlebars = require('handlebars');
var moment = require('moment');
var encode = require('email-scramble').encode;

function setVar(name, value, options) {
  options.data.root[name] = value;
}

function formatDate(date, format) {
  return new Handlebars.SafeString(moment(Handlebars.escapeExpression(date || '')).format(format));
}

function now() {
  return new Date();
}

function year() {
  return now().getFullYear();
}

function toLowerCase(s) {
  return (s || '').toLowerCase();
}

function isUrl(str, options) {
  var regex = /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g;
  return regex.test(str || '') ? options.fn(this) : options.inverse(this);
}

function isEmail(str, options) {
  var regex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return regex.test(str || '') ? options.fn(this) : options.inverse(this);
}

function replace(haystack, needle, substitute) {
  return haystack.replace(needle, substitute);
}

module.exports =  {
  register: function() {
    Handlebars.registerHelper({
      setVar,
      formatDate,
      year,
      now,
      encode,
      toLowerCase,
      isUrl,
      isEmail,
      replace
    })
  }
};
