// jsonresume-theme-gnap/index.js
var fs = require("fs");
var Handlebars = require("handlebars");
var helpers = require('./helpers');
var partials = require('./partials');

const defaultHeaders = {
  "basics": "Profile",
  "skills": "Skills",
  "languages": "Languages",
  "projects": "Projects",
  "work": "Work",
  "education": "Education",
  "references": "References",
  "interests": "Interests",
  "certificates": "Certificates",
  "awards": "Awards",
  "volunteer": "Volunteer",
  "publications": "Publications",
  "contactForm": "Contact"
};

function withDefaults(resume) {
  resume.meta = {
    ...resume.meta
  };
  resume.meta.headers = {
    ...defaultHeaders,
    ...resume.meta.headers
  };

  if (resume.meta && resume.meta.contactForm && undefined === resume.meta.contactForm.header) {
    resume.meta.contactForm = {
      header: defaultHeaders.contactForm,
      ...resume.meta.contactForm
    };
  }

  return { ...resume };
}

function render(resume) {
  const template = fs.readFileSync(__dirname + '/gnap.hbs', 'utf-8');
  const options = {
    css: fs.readFileSync(__dirname + '/gnap.css', 'utf-8'),
    js: fs.readFileSync(__dirname + '/gnap.js', 'utf-8'),
    resume: withDefaults(resume)
  };

  helpers.register();
  partials.register();

  return Handlebars.compile(template)(options);
}


module.exports = { render, defaultHeaders };
